File: hotel-api/src/main/java/br/com/luizgustavo/hotelapi/utils/PriceCalculator.java
package br.com.luizgustavo.hotelapi.utils;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalTime;

import br.com.luizgustavo.hotelapi.model.Booking;

public class PriceCalculator {
	
	private static final Double NORMALPRICE = 120.0;
	private static final Double WEEKENDPRICE = 150.0;
	private static final Double NORMALPARKINGPRICE = 15.0;
	private static final Double WEEKENDPARKINGPRICE = 20.0;
	private static final LocalTime LIMITHOURCHECKOUT = LocalTime.of(16, 30);
			
	public static void calculateBookingPrice(Booking booking) {
		LocalDate date = booking.getCheckIn().toLocalDate();
		
		double price = 0.0;
		while (date.isBefore(booking.getCheckOut().toLocalDate())) {
			price += calculatePriceOfDate(date, booking.useParking());			
			date = date.plusDays(1);						
		}

		if (booking.getHourCheckOut().isAfter(LIMITHOURCHECKOUT)) {
			price += calculatePriceOfDate(booking.getCheckOut().plusDays(1).toLocalDate(), booking.useParking());			
		}

		booking.setPrice(price);
	}
	
	private static double calculatePriceOfDate(LocalDate date, boolean useParking) {
		double price = 0.0;
		price += isWeekend(date) ? WEEKENDPRICE : NORMALPRICE;
		if (useParking) {
			price += isWeekend(date) ? WEEKENDPARKINGPRICE : NORMALPARKINGPRICE;
		}
		return price;
	}
	
	private static boolean isWeekend(LocalDate date) {
		DayOfWeek day = date.getDayOfWeek();
		return day.equals(DayOfWeek.SATURDAY) || day.equals(DayOfWeek.SUNDAY);		
	}

}

File: hotel-front/src/app/booking/price-calculator.ts
import * as moment from 'moment';

export class PriceCalculator {
  
  private static NORMALPRICE: number = 120.0;
  private static WEEKENDPRICE: number = 150.0;
  private static NORMALPARKINGPRICE: number = 15.0;
  private static WEEKENDPARKINGPRICE: number = 20.0;
  
  static calculateBookingPrice(checkIn: Date, checkOut: Date, useParking: boolean): number {
    const dateCheckout = moment(checkOut, 'YYYY-MM-DD');
    let date = moment(checkIn, 'YYYY-MM-DD');
    
    let price = 0;    
    while (date.isBefore(dateCheckout)) {      
      price += this.calculatePriceOfDate(date, useParking);
      date = moment(date).add(1, 'days');
    }
    
    if (this.newDaily(checkOut)) {
      price += this.calculatePriceOfDate(moment(date).add(1, 'days'), useParking);      
    }
    
    return price;
  }

  private static calculatePriceOfDate(data: moment.Moment, adicionalVeiculo: boolean): number {
    let price = 0;
    price += this.isWeekend(data) ? this.WEEKENDPRICE : this.NORMALPRICE;
    if (adicionalVeiculo) {
      price += this.isWeekend(data) ? this.WEEKENDPARKINGPRICE : this.NORMALPARKINGPRICE;
    }
    return price;
  } 

  private static isWeekend(data: moment.Moment): boolean {
    return data.weekday() === 0 || data.weekday() === 6;
  }

  private static newDaily(checkOut: Date) {
    const dia = parseInt(moment(checkOut).format('DD'));
    const mes = parseInt(moment(checkOut).format('MM')) - 1;
    const ano = parseInt(moment(checkOut).format('YYYY'));

    const horaLimite = moment(new Date(ano, mes, dia, 16, 30));
    const horaCheckout = moment(checkOut);

    return horaCheckout.isAfter(horaLimite);
  }
}